### ISSUE TITLE
permanently hanging workers

### ISSUE THREAD

**Issue by jfmyers9** [concourse/concourse#1035]
# Bug Report

This bug report has 2 issues:
- Defunct runc processes from baggage claim (probably not the root cause?)
- Hanging workers due to btrfs issues with baggage claim (probably the root cause?)

Here is some basic information:
* Concourse version: 2.7.3
* Deployment type: BOSH
* Infrastructure: GCP
* Did this used to work?: Yes.

Starting randomly on April 29th all of our concourse builds began to hang indefinitely. Here is the output from `bosh vms --vitals`:
```
inigo-worker-mysql/7475bf21-a0cb-44fc-a056-ad740947f523             running        z1  10.0.16.7   vm-5f939a84-9b18-4f95-6e1d-8792fbb510fd  n1-standard-4  -       0.00, 0.00, 0.00        -      0.0%  0.0%  0.0%  3% (538 MB)  0% (0 B)     43% (33i%)  2% (1i%)    -
inigo-worker-postgres-grootfs/e9019ce4-cd78-42df-8daa-447cb90cd374  running        z1  10.0.16.13  vm-f717a014-a8ec-4a03-64b1-dd88d5bb7b45  n1-standard-4  -       0.03, 0.01, 0.00        -      0.0%  0.0%  0.0%  3% (398 MB)  0% (6.2 MB)  43% (33i%)  2% (1i%)    -
inigo-worker-postgres/5524338f-6025-4f1d-9c84-dfbf458f585e          running        z1  10.0.16.8   vm-369214a3-1dcf-44df-5f04-9e3315eacab4  n1-standard-4  -       0.00, 0.00, 0.00        -      0.0%  0.0%  0.0%  4% (555 MB)  0% (0 B)     43% (33i%)  2% (1i%)    -
postgresql/d8d5c929-929a-488d-a8ff-3de8e50e3acc                     running        z1  10.0.16.6   vm-2ddd7cac-7131-401b-6ae1-a3f55bfdcca7  m3.large       -       0.20, 0.16, 0.16        -      5.4%  1.3%  1.3%  3% (206 MB)  0% (0 B)     43% (33i%)  2% (1i%)    5% (0i%)
web/07638cad-07ab-46e8-a52f-0ee3ff75989d                            running        z1  10.0.16.4   vm-66277759-3f3c-4438-6658-f93987b59878  m3.medium      -       0.00, 0.00, 0.00        -      2.6%  0.7%  0.0%  6% (222 MB)  0% (0 B)     43% (33i%)  1% (0i%)    -
web/d3bc2f74-3aaa-42f1-a8f3-3e5f19ff62bf                            running        z1  10.0.16.5   vm-8bc1780e-3bb9-4073-6781-b27d371ce57d  m3.medium      -       0.03, 0.03, 0.00        -      4.0%  1.3%  0.0%  4% (150 MB)  0% (0 B)     43% (33i%)  1% (0i%)    -
worker/62783da3-b4eb-4b46-9d42-dd896d177732                         running        z1  10.0.16.9   vm-90ea40f5-9e16-443f-7638-d6d459b9f153  n1-standard-8  -       0.06, 0.09, 0.10        -      1.8%  1.2%  0.1%  6% (1.9 GB)  1% (188 MB)  43% (33i%)  12% (4i%)   -
worker/760effa6-1d6f-4e4b-8aff-df6b328d79cf                         running        z1  10.0.16.10  vm-b22a2de9-ec93-4f41-524a-62db2ed36b68  n1-standard-8  -       0.13, 0.10, 0.09        -      0.4%  0.4%  0.0%  7% (2.3 GB)  1% (160 MB)  43% (33i%)  11% (4i%)   -
worker/83f975df-0f82-4117-8217-6255fe39fe38                         running        z1  10.0.16.11  vm-e9149135-c92f-4893-4b1d-8956e836cd24  n1-standard-8  -       0.03, 0.07, 0.08        -      2.3%  0.8%  0.0%  7% (2.1 GB)  1% (165 MB)  43% (33i%)  12% (5i%)   -
worker/8a1baf02-41b4-4ff4-8473-44a4ebf0c673                         running        z1  10.0.16.12  vm-a2e0cc82-b79e-44d0-5193-d267696b3987  n1-standard-8  -       170.03, 169.71, 165.24  -      0.2%  0.3%  0.0%  7% (2.3 GB)  0% (133 MB)  43% (33i%)  12% (5i%)   -
```

The weirdest thing here was the final worker with ridiculously high usage. After sshing onto the vm I found that there were countless defunct runc processes:
```
root       18947  0.0  0.0      0     0 ?        D<s  Apr29   0:00 [runc:[2:INIT]]
root       18955  0.0  0.0 826860 15008 ?        S<l  Apr29   0:44 /var/vcap/packages/runc/bin/runc events b3eac29a-11f9-40e9-55c9-565e5acbf1f0
root       20438  0.0  0.0      0     0 ?        D<s  Apr29   0:00 [runc:[2:INIT]]
root       20446  0.0  0.0 687592 17264 ?        S<l  Apr29   0:41 /var/vcap/packages/runc/bin/runc events 9664cd9d-6d84-4f16-4995-f6f58b9ffdcb
root       28208  0.0  0.0      0     0 ?        Z<sl Apr29   0:00 [runc:[2:INIT]] <defunct>
root       28216  0.0  0.0 630508 14956 ?        S<l  Apr29   0:38 /var/vcap/packages/runc/bin/runc events 8262f98b-638e-4722-5a25-de83f1032042
root       32743  0.0  0.0      0     0 ?        D<s  Apr29   0:00 [runc:[2:INIT]]
root       32751  0.0  0.0 826860 16964 ?        S<l  Apr29   0:42 /var/vcap/packages/runc/bin/runc events c3c0b59a-4567-4f03-5bd9-f6472bb0a4cf
root       39462  0.0  0.0      0     0 ?        Z<sl Apr29   0:00 [runc:[2:INIT]] <defunct>
```

The output from top sorted by defunct processes:
```
top - 18:25:22 up 13 days, 15:17,  1 user,  load average: 170.00, 170.00, 168.03
Tasks: 294 total,   1 running, 285 sleeping,   0 stopped,   8 zombie
%Cpu(s):  0.2 us,  0.2 sy,  0.0 ni, 99.6 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem:  30882916 total, 27690056 used,  3192860 free,   237604 buffers
KiB Swap: 30882912 total,   132416 used, 30750496 free. 25135536 cached Mem

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
   5689 root      10 -10 5224752  32920   3900 D   0.0  0.1 177:23.03 baggageclaim
   5771 root      20   0       0      0      0 D   0.0  0.0   9:09.28 btrfs-transacti
  18947 root      10 -10       0      0      0 D   0.0  0.0   0:00.15 runc:[2:INIT]
  20438 root      10 -10       0      0      0 D   0.0  0.0   0:00.16 runc:[2:INIT]
  32743 root      10 -10       0      0      0 D   0.0  0.0   0:00.15 runc:[2:INIT]
  49154 root      10 -10       0      0      0 D   0.0  0.0   0:00.17 runc:[2:INIT]
  56166 root      10 -10       0      0      0 D   0.0  0.0   0:00.17 runc:[2:INIT]
  56426 root      10 -10       0      0      0 D   0.0  0.0   0:00.16 runc:[2:INIT]
  57513 root      10 -10       0      0      0 D   0.0  0.0   0:00.15 runc:[2:INIT]
  60649 root      10 -10       0      0      0 D   0.0  0.0   0:00.16 runc:[2:INIT]
  95961 root      10 -10       0      0      0 D   0.0  0.0   0:00.16 runc:[2:INIT]
 102907 root      10 -10       0      0      0 D   0.0  0.0   0:00.15 runc:[2:INIT]
 118193 root      10 -10       0      0      0 D   0.0  0.0   0:00.15 runc:[2:INIT]
 133515 root      20   0       0      0      0 D   0.0  0.0   0:02.48 kworker/u16:8
 133664 root      10 -10       0      0      0 D   0.0  0.0   0:00.14 runc:[2:INIT]
```

Then looking into these defunct processes. The output from `ps axo pid,ppid,comm,state,eip,esp,wchan | grep 'D\|Z'`:
```
    PID    PPID COMMAND         S      EIP      ESP WCHAN
   5689       1 baggageclaim    D 00478454 2061d590 filename_create
   5771       2 btrfs-transacti D 00000000 00000000 wait_current_trans.isra.24
  18947       1 runc:[2:INIT]   D 00000000 00000000 zap_pid_ns_processes
  20438       1 runc:[2:INIT]   D 00000000 00000000 zap_pid_ns_processes
  28208       1 runc: <defunct> Z 00000000 00000000 exit
  32743       1 runc:[2:INIT]   D 00000000 00000000 zap_pid_ns_processes
  39462       1 runc: <defunct> Z 00000000 00000000 exit
  49154       1 runc:[2:INIT]   D 00000000 00000000 zap_pid_ns_processes
  53208       1 runc: <defunct> Z 00000000 00000000 exit
  56166       1 runc:[2:INIT]   D 00000000 00000000 zap_pid_ns_processes
  56426       1 runc:[2:INIT]   D 00000000 00000000 zap_pid_ns_processes
  57513       1 runc:[2:INIT]   D 00000000 00000000 zap_pid_ns_processes
  60649       1 runc:[2:INIT]   D 00000000 00000000 zap_pid_ns_processes
  61371       1 runc: <defunct> Z 00000000 00000000 exit
  83917       1 runc: <defunct> Z 00000000 00000000 exit
  88156       1 runc: <defunct> Z 00000000 00000000 exit
  95961       1 runc:[2:INIT]   D 00000000 00000000 zap_pid_ns_processes
 102907       1 runc:[2:INIT]   D 00000000 00000000 zap_pid_ns_processes
 118193       1 runc:[2:INIT]   D 00000000 00000000 zap_pid_ns_processes
 125458       1 runc: <defunct> Z 00000000 00000000 exit
 133515       2 kworker/u16:8   D 00000000 00000000 wait_current_trans.isra.24
 133664       1 runc:[2:INIT]   D 00000000 00000000 zap_pid_ns_processes
 137451  136681 dpkg            D 901b9b80 a23bda18 wait_current_trans.isra.24
 137995  137851 dockerd         D 00694554 20edd5f8 unlinkat
 138022  137995 docke <defunct> Z 00000000 00000000 exit
 138034  136243 dpkg            D 53da6b80 c58072f8 btrfs_commit_transaction
 138788       1 git             D 2437cd97 3dc68e68 wait_current_trans.isra.24
 138877       1 check           D 25501080 5426fcf8 wait_current_trans.isra.24
 138905  138887 check           D 9db00f17 7131f308 wait_current_trans.isra.24
 139010       1 check           D 38a01080 6e958e48 wait_current_trans.isra.24
 139099       1 check           D b430d080 32522d78 wait_current_trans.isra.24
 139136       1 check           D cbc82080 85169af8 wait_current_trans.isra.24
 139174       1 check           D 30690080 987392b8 wait_current_trans.isra.24
 139207  139189 check           D f897cf17 6f1910f8 wait_current_trans.isra.24
 139292       1 check           D 02b6d080 b259aab8 wait_current_trans.isra.24
 139331       1 check           D 9e5e1080 b0580658 wait_current_trans.isra.24
 139372       1 check           D c9f4e080 8dd86d28 wait_current_trans.isra.24
 139463       1 check           D 987f7080 e80fa2f8 wait_current_trans.isra.24
 139567  139549 check           D 561bcf17 e8eb62a8 wait_current_trans.isra.24
 139613  139595 check           D 5e66af17 6b8faa38 wait_current_trans.isra.24
 139651  139633 check           D f1be1f17 e633d318 wait_current_trans.isra.24
 139705       1 check           D 5d5ff080 1b30dc38 wait_current_trans.isra.24
 139735       1 check           D 10f8e080 b071e838 wait_current_trans.isra.24
 139841       1 check           D 38706080 a85fded8 wait_current_trans.isra.24
```
From the above it appears that a bunch of the runc defunct processes may be waiting on a btrfs transaction committing, which is also hanging. Also baggage claim itself is hung on `filename_create`, which is weird.

This appears to be similar to the grootfs issue surrounding defunct processes. Maybe baggage claim suffers from a similar bug. Which seems worrying? Can one hanging worker cause the ATC to break?

I don't think so, so I recreated the worker, and noticed that builds were still hanging indefinitely. Following some more investigation, I noticed that builds that can't possibly run on this worker were also hanging. Putting in some further investigation, one of the first builds to fail had logs like:
```
{"timestamp":"1493530243.697216511","source":"atc","message":"atc.container-keepaliver.looking-up-container","log_level":1,"data":{"handle":"1f9a076b-69df-4394-4f0e-0bfdab5d4b78","session":"22"}}
{"timestamp":"1493532164.216517687","source":"atc","message":"atc.baggage-collector.could-not-locate-worker","log_level":1,"data":{"session":"18","worker-id":"8a1baf02-e326-4341-ab00-4f94b48054e1"}}
{"timestamp":"1493532344.111821651","source":"atc","message":"atc.baggage-collector.could-not-locate-worker","log_level":1,"data":{"session":"18","worker-id":"8a1baf02-e326-4341-ab00-4f94b48054e1"}}
{"timestamp":"1493533694.118216515","source":"atc","message":"atc.baggage-collector.could-not-locate-worker","log_level":1,"data":{"session":"18","worker-id":"8a1baf02-e326-4341-ab00-4f94b48054e1"}}
{"timestamp":"1493534084.182165146","source":"atc","message":"atc.baggage-collector.could-not-locate-worker","log_level":1,"data":{"session":"18","worker-id":"8a1baf02-e326-4341-ab00-4f94b48054e1"}}
{"timestamp":"1493536214.121651649","source":"atc","message":"atc.baggage-collector.could-not-locate-worker","log_level":1,"data":{"session":"18","worker-id":"62783da3-ba67-4132-8012-25c1919d4879"}}
{"timestamp":"1493540054.212165117","source":"atc","message":"atc.container-keepaliver.looking-up-container","log_level":1,"data":{"handle":"f59ebbed-bde1-4634-4101-5d9666e87ffa","session":"22"}}
{"timestamp":"1493548244.216516495","source":"atc","message":"atc.container-keepaliver.looking-up-container","log_level":1,"data":{"handle":"161d6b44-716e-4688-786f-770289fdb99f","session":"22"}}
{"timestamp":"1493548963.942165136","source":"atc","message":"atc.container-keepaliver.volume.initial-heartbeat.failed-to-get-volume-size","log_level":2,"data":{"container":"0bdd48dc-d089-4e66-4dc0-593f2d9caede","error":"failed to get volume stats","handle":"18dc48d4-fc11-44fe-40d3-bccd5197585c","session":"22.38.1","volume":"18dc48d4-fc11-44fe-40d3-bccd5197585c"}}
```

 At this point the logs for the above build had rotated for this build but any new build would fail similarly:
```
{"timestamp":"1493652464.216677427","source":"atc","message":"atc.baggage-collector.could-not-locate-worker","log_level":1,"data":{"session":"18","worker-id":"8a1baf02-e326-4341-ab00-4f94b48054e1"}}
{"timestamp":"1493652584.216676712","source":"atc","message":"atc.baggage-collector.could-not-locate-worker","log_level":1,"data":{"session":"18","worker-id":"62783da3-ba67-4132-8012-25c1919d4879"}}
{"timestamp":"1493653184.233321667","source":"atc","message":"atc.baggage-collector.could-not-locate-worker","log_level":1,"data":{"session":"18","worker-id":"83f975df-5495-422b-9545-d283484f445e"}}
{"timestamp":"1493654593.762166739","source":"atc","message":"atc.container-keepaliver.volume.initial-heartbeat.failed-to-get-volume-size","log_level":2,"data":{"container":"838901ae-82c9-4e76-72d1-40f2964e45bb","error":"failed to get volume stats","handle":"cbf93b03-4408-4e8d-5723-8e902f9ca112","session":"22.24.1","volume":"cbf93b03-4408-4e8d-5723-8e902f9ca112"}}
{"timestamp":"1493656214.129216671","source":"atc","message":"atc.baggage-collector.could-not-locate-worker","log_level":1,"data":{"session":"18","worker-id":"8a1baf02-e326-4341-ab00-4f94b48054e1"}}
{"timestamp":"1493658417.216675043","source":"atc","message":"atc.container-keepaliver.volume.initial-heartbeat.failed-to-get-volume-size","log_level":2,"data":{"container":"a68a9b5c-66e3-4bd3-6625-14465a03219a","error":"failed to get volume stats","handle":"393d278e-32f1-4e19-605a-7a84640e6cfe","session":"22.38.1","volume":"393d278e-32f1-4e19-605a-7a84640e6cfe"}}
{"timestamp":"1493659317.216674805","source":"atc","message":"atc.baggage-collector.could-not-locate-worker","log_level":1,"data":{"session":"18","worker-id":"83f975df-5495-422b-9545-d283484f445e"}}
{"timestamp":"1493659887.175216675","source":"atc","message":"atc.baggage-collector.could-not-locate-worker","log_level":1,"data":{"session":"18","worker-id":"83f975df-5495-422b-9545-d283484f445e"}}
{"timestamp":"1493659945.704348564","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.scan.init.volume.initial-heartbeat.failed-to-get-volume-size","log_level":2,"data":{"build-id":21667,"build-name":"314","container":"8c075e21-078a-421f-4588-07b2810891c2","error":"failed to get volume stats","handle":"b61886dc-aad4-424a-6d67-bb1ef8f1fd36","input":"diego-release","job_name":"inigo-mysql","resource":"diego-release","session":"14748.1.1.2.3.3.1","volume":"b61886dc-aad4-424a-6d67-bb1ef8f1fd36"}}
{"timestamp":"1493659946.471603632","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.scan.init.volume.initial-heartbeat.failed-to-get-volume-size","log_level":2,"data":{"build-id":21667,"build-name":"314","container":"f990f134-5cf7-4551-664e-fc5b1e5a4255","error":"failed to get volume stats","handle":"393f2a40-868e-4af8-562e-fb89ad73f9e2","input":"garden-runc-release-tarball","job_name":"inigo-mysql","resource":"garden-runc-release-tarball","session":"14748.1.1.3.3.3.1","volume":"393f2a40-868e-4af8-562e-fb89ad73f9e2"}}
{"timestamp":"1493659947.180071115","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.scan.init.volume.initial-heartbeat.failed-to-get-volume-size","log_level":2,"data":{"build-id":21667,"build-name":"314","container":"254e1768-620a-426f-67c9-ddc2a4a052fa","error":"failed to get volume stats","handle":"50b7eec5-d256-4f98-4a62-27b6689b375c","input":"diego-ci","job_name":"inigo-mysql","resource":"diego-ci","session":"14748.1.1.5.3.3.1","volume":"50b7eec5-d256-4f98-4a62-27b6689b375c"}}
{"timestamp":"1493659948.274888754","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.starting","log_level":1,"data":{"build-id":21667,"build-name":"314","job_name":"inigo-mysql","session":"14748.1.1"}}
{"timestamp":"1493659948.280337095","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.running","log_level":1,"data":{"build":21667,"build-id":21667,"build-name":"314","job":"inigo-mysql","job_name":"inigo-mysql","pipeline":"main","session":"14748.1.1"}}
{"timestamp":"1493659948.296042204","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.do.aggregate.get.lock-task.tick","log_level":1,"data":{"build-id":21667,"build-name":"314","job_name":"inigo-mysql","lock-name":"31d8bd7829da081852a11c512dd1c09e0817c10ddfceeef4113a61468e29b8ed","name":"diego-release","session":"14748.1.1.9.3.1.3"}}
{"timestamp":"1493659948.311697006","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.do.aggregate.get.initial-heartbeat.failed-to-get-volume-size","log_level":2,"data":{"build-id":21667,"build-name":"314","error":"failed to get volume stats","job_name":"inigo-mysql","name":"diego-ci","session":"14748.1.1.9.3.4.1","volume":"e646c660-56ad-4ffe-63ba-84f524babeff"}}
{"timestamp":"1493659948.316114426","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.do.aggregate.get.initial-heartbeat.failed-to-get-volume-size","log_level":2,"data":{"build-id":21667,"build-name":"314","error":"failed to get volume stats","job_name":"inigo-mysql","name":"garden-runc-release-tarball","session":"14748.1.1.9.3.2.1","volume":"6000941c-51c5-4e05-6394-6584449c5dd1"}}
{"timestamp":"1493659948.334841490","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.do.aggregate.get.finished","log_level":1,"data":{"build-id":21667,"build-name":"314","job_name":"inigo-mysql","name":"diego-ci","session":"14748.1.1.9.3.4","version-info":{"Version":{"ref":"93676253d7aa9a618bec90a3de45f70c273eb4ad"},"Metadata":null}}}
{"timestamp":"1493659948.338879585","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.do.aggregate.get.finished","log_level":1,"data":{"build-id":21667,"build-name":"314","job_name":"inigo-mysql","name":"garden-runc-release-tarball","session":"14748.1.1.9.3.2","version-info":{"Version":{"version":"1.5.0"},"Metadata":null}}}
{"timestamp":"1493659978.325373888","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.do.aggregate.get.continuous-heartbeat.tick.failed-to-get-volume-size","log_level":2,"data":{"build-id":21667,"build-name":"314","error":"failed to get volume stats","job_name":"inigo-mysql","name":"diego-ci","session":"14748.1.1.9.3.4.2.1","volume":"e646c660-56ad-4ffe-63ba-84f524babeff"}}
{"timestamp":"1493660008.325428963","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.do.aggregate.get.continuous-heartbeat.tick.failed-to-get-volume-size","log_level":2,"data":{"build-id":21667,"build-name":"314","error":"failed to get volume stats","job_name":"inigo-mysql","name":"diego-ci","session":"14748.1.1.9.3.4.2.2","volume":"e646c660-56ad-4ffe-63ba-84f524babeff"}}
{"timestamp":"1493660012.719087124","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.do.aggregate.get.finished","log_level":1,"data":{"build-id":21667,"build-name":"314","job_name":"inigo-mysql","name":"diego-release","session":"14748.1.1.9.3.1","version-info":{"Version":{"ref":"2dff8f5b3573d32609e0d116efcdd8d502733f7a"},"Metadata":[{"name":"commit","value":"2dff8f5b3573d32609e0d116efcdd8d502733f7a"},{"name":"author","value":"James Myers"},{"name":"author_date","value":"2017-04-28 17:06:05 -0700"},{"name":"branch","value":"develop"},{"name":"message","value":"bump executor\n\n[finishes #144561585]\n\nSubmodule src/code.cloudfoundry.org/executor 789c2f8b..91aa1b0e:\n  \u003e fix flaky reporter_test\n\nSigned-off-by: Chris Piraino \u003ccpiraino@pivotal.io\u003e\n"}]}}}
{"timestamp":"1493660038.325417519","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.do.aggregate.get.continuous-heartbeat.tick.failed-to-get-volume-size","log_level":2,"data":{"build-id":21667,"build-name":"314","error":"failed to get volume stats","job_name":"inigo-mysql","name":"diego-ci","session":"14748.1.1.9.3.4.2.3","volume":"e646c660-56ad-4ffe-63ba-84f524babeff"}}
{"timestamp":"1493660068.325038671","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.do.aggregate.get.continuous-heartbeat.tick.failed-to-get-volume-size","log_level":2,"data":{"build-id":21667,"build-name":"314","error":"failed to get volume stats","job_name":"inigo-mysql","name":"diego-ci","session":"14748.1.1.9.3.4.2.4","volume":"e646c660-56ad-4ffe-63ba-84f524babeff"}}
{"timestamp":"1493660098.325416565","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.do.aggregate.get.continuous-heartbeat.tick.failed-to-get-volume-size","log_level":2,"data":{"build-id":21667,"build-name":"314","error":"failed to get volume stats","job_name":"inigo-mysql","name":"diego-ci","session":"14748.1.1.9.3.4.2.5","volume":"e646c660-56ad-4ffe-63ba-84f524babeff"}}
{"timestamp":"1493660128.318360090","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.do.aggregate.get.continuous-heartbeat.tick.failed-to-get-volume-size","log_level":2,"data":{"build-id":21667,"build-name":"314","error":"failed to get volume stats","job_name":"inigo-mysql","name":"diego-ci","session":"14748.1.1.9.3.4.2.6","volume":"e646c660-56ad-4ffe-63ba-84f524babeff"}}
{"timestamp":"1493660158.325431585","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.do.aggregate.get.continuous-heartbeat.tick.failed-to-get-volume-size","log_level":2,"data":{"build-id":21667,"build-name":"314","error":"failed to get volume stats","job_name":"inigo-mysql","name":"diego-ci","session":"14748.1.1.9.3.4.2.7","volume":"e646c660-56ad-4ffe-63ba-84f524babeff"}}
{"timestamp":"1493660188.332024574","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.do.aggregate.get.continuous-heartbeat.tick.failed-to-get-volume-size","log_level":2,"data":{"build-id":21667,"build-name":"314","error":"failed to get volume stats","job_name":"inigo-mysql","name":"diego-ci","session":"14748.1.1.9.3.4.2.8","volume":"e646c660-56ad-4ffe-63ba-84f524babeff"}}
{"timestamp":"1493660218.325583220","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.do.aggregate.get.continuous-heartbeat.tick.failed-to-get-volume-size","log_level":2,"data":{"build-id":21667,"build-name":"314","error":"failed to get volume stats","job_name":"inigo-mysql","name":"diego-ci","session":"14748.1.1.9.3.4.2.9","volume":"e646c660-56ad-4ffe-63ba-84f524babeff"}}
{"timestamp":"1493660248.325741291","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.do.aggregate.get.continuous-heartbeat.tick.failed-to-get-volume-size","log_level":2,"data":{"build-id":21667,"build-name":"314","error":"failed to get volume stats","job_name":"inigo-mysql","name":"diego-ci","session":"14748.1.1.9.3.4.2.10","volume":"e646c660-56ad-4ffe-63ba-84f524babeff"}}
{"timestamp":"1493660278.325653791","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.do.aggregate.get.continuous-heartbeat.tick.failed-to-get-volume-size","log_level":2,"data":{"build-id":21667,"build-name":"314","error":"failed to get volume stats","job_name":"inigo-mysql","name":"diego-ci","session":"14748.1.1.9.3.4.2.11","volume":"e646c660-56ad-4ffe-63ba-84f524babeff"}}
```

Following the paper trail I ended up in the baggage claim where I found logs that look like this:
```
{"timestamp":"1493658657.725124598","source":"baggageclaim","message":"baggageclaim.driver.run-command.failed","log_level":2,"data":{"args":["qgroup","show","-F","--raw","/var/vcap/data/baggageclaim/volumes/live/6b635d05-dba0-490d-6c74-3375e9b8ee5e/volume"],"command":"/var/vcap/packages/btrfs_tools/sbin/btrfs","error":"exit status 1","session":"1.2490402","stderr":"ERROR: can't perform the search - No such file or directory\nERROR: can't list qgroups: No such file or directory\n","stdout":""}}
{"timestamp":"1493658657.725192785","source":"baggageclaim","message":"baggageclaim.repository.get-volume.failed-to-get-volume-stats","log_level":2,"data":{"error":"exit status 1","session":"3.21863404","volume":"6b635d05-dba0-490d-6c74-3375e9b8ee5e"}}
{"timestamp":"1493658657.979275703","source":"baggageclaim","message":"baggageclaim.driver.run-command.failed","log_level":2,"data":{"args":["qgroup","show","-F","--raw","/var/vcap/data/baggageclaim/volumes/live/7ec0c61f-c354-450c-4508-3e13d6cbdf70/volume"],"command":"/var/vcap/packages/btrfs_tools/sbin/btrfs","error":"exit status 1","session":"1.2490403","stderr":"ERROR: can't perform the search - No such file or directory\nERROR: can't list qgroups: No such file or directory\n","stdout":""}}
{"timestamp":"1493658657.979340315","source":"baggageclaim","message":"baggageclaim.repository.get-volume.failed-to-get-volume-stats","log_level":2,"data":{"error":"exit status 1","session":"3.21863408","volume":"7ec0c61f-c354-450c-4508-3e13d6cbdf70"}}
{"timestamp":"1493658658.057608366","source":"baggageclaim","message":"baggageclaim.driver.run-command.failed","log_level":2,"data":{"args":["qgroup","show","-F","--raw","/var/vcap/data/baggageclaim/volumes/live/e245e30b-db47-4c63-4ad4-3a4bd377c49c/volume"],"command":"/var/vcap/packages/btrfs_tools/sbin/btrfs","error":"exit status 1","session":"1.2490404","stderr":"ERROR: can't perform the search - No such file or directory\nERROR: can't list qgroups: No such file or directory\n","stdout":""}}
{"timestamp":"1493658658.057688713","source":"baggageclaim","message":"baggageclaim.repository.get-volume.failed-to-get-volume-stats","log_level":2,"data":{"error":"exit status 1","session":"3.21863412","volume":"e245e30b-db47-4c63-4ad4-3a4bd377c49c"}}
{"timestamp":"1493658662.267828226","source":"baggageclaim","message":"baggageclaim.driver.run-command.failed","log_level":2,"data":{"args":["qgroup","show","-F","--raw","/var/vcap/data/baggageclaim/volumes/live/fa2e84d5-8c42-4594-5ac6-2201482bcb5e/volume"],"command":"/var/vcap/packages/btrfs_tools/sbin/btrfs","error":"exit status 1","session":"1.2490405","stderr":"ERROR: can't perform the search - No such file or directory\nERROR: can't list qgroups: No such file or directory\n","stdout":""}}
{"timestamp":"1493658662.267899275","source":"baggageclaim","message":"baggageclaim.repository.get-volume.failed-to-get-volume-stats","log_level":2,"data":{"error":"exit status 1","session":"3.21863551","volume":"fa2e84d5-8c42-4594-5ac6-2201482bcb5e"}}
{"timestamp":"1493658662.648453712","source":"baggageclaim","message":"baggageclaim.driver.run-command.failed","log_level":2,"data":{"args":["qgroup","show","-F","--raw","/var/vcap/data/baggageclaim/volumes/live/a9557d4c-8bf6-4ebf-69b5-0c2ab2976210/volume"],"command":"/var/vcap/packages/btrfs_tools/sbin/btrfs","error":"exit status 1","session":"1.2490406","stderr":"ERROR: can't perform the search - No such file or directory\nERROR: can't list qgroups: No such file or directory\n","stdout":""}}
{"timestamp":"1493658662.648555994","source":"baggageclaim","message":"baggageclaim.repository.get-volume.failed-to-get-volume-stats","log_level":2,"data":{"error":"exit status 1","session":"3.21863553","volume":"a9557d4c-8bf6-4ebf-69b5-0c2ab2976210"}}
{"timestamp":"1493658663.082960367","source":"baggageclaim","message":"baggageclaim.driver.run-command.failed","log_level":2,"data":{"args":["qgroup","show","-F","--raw","/var/vcap/data/baggageclaim/volumes/live/c5a087d4-fa24-48e0-4702-dabda6fe6031/volume"],"command":"/var/vcap/packages/btrfs_tools/sbin/btrfs","error":"exit status 1","session":"1.2490407","stderr":"ERROR: can't perform the search - No such file or directory\nERROR: can't list qgroups: No such file or directory\n","stdout":""}}
{"timestamp":"1493658663.083039999","source":"baggageclaim","message":"baggageclaim.repository.get-volume.failed-to-get-volume-stats","log_level":2,"data":{"error":"exit status 1","session":"3.21863557","volume":"c5a087d4-fa24-48e0-4702-dabda6fe6031"}}
{"timestamp":"1493658665.845414639","source":"baggageclaim","message":"baggageclaim.driver.run-command.failed","log_level":2,"data":{"args":["qgroup","show","-F","--raw","/var/vcap/data/baggageclaim/volumes/live/d54263ae-d055-42ed-7b63-24dbc900f11d/volume"],"command":"/var/vcap/packages/btrfs_tools/sbin/btrfs","error":"exit status 1","session":"1.2490408","stderr":"ERROR: can't perform the search - No such file or directory\nERROR: can't list qgroups: No such file or directory\n","stdout":""}}
```

I'm sure there are some other errors in there, but the logs are such high volume that it is really difficult to parse. We see similar issues on all of our workers, and I'm not sure what is a red herring and what is an issue. But there does seem to be something wrong within the realm of BaggageClaim + ATC + RunC + Btrfs.

Some interesting things:
* This happened randomly with no user interaction (a deploy, bosh recreate, etc) after 13+ days of operating on 2.7.3.
* Hanging builds that were cancelled were unable to leave the pending state.
* This affected all workers. (How is that even possible?) We have 4 types workers that are responsible for running different work loads. They are all exhibiting this behavior.
* `bosh recreate` on the worker with runc defunct processes hung forever. I needed to delete the VM from the gcp console.
* Recreating the worker with defunct processes did not unblock concourse jobs. (Which leads me to believe that this isn't the root cause?)
* **After doing a `bosh -d concourse recreate`, jobs were able to trigger again and the pending jobs were unblocked.**
* I'm very confused.

**Comment by joshzarrabi** [comment]
@jfmyers9 Thanks for taking the time to submit such a detailed bug report...
We have seen some other issues with btrfs hanging and causing havoc in our own instance, so we have done some work to alleviate that with the next release, which you can see here https://www.pivotaltracker.com/story/show/142065347. 

It seems like sometimes btrfs processes hang forever, which confuses baggageclaim and thus atc. I am a little bit surprised that bosh recreating your worker does not fix this though...
What is the output when you run `fly workers`? Are any of your workers stalled?

**Comment by vito** [comment]
I think this is the same as https://github.com/concourse/concourse/issues/767 - stay tuned for an upcoming issue proposing we find an alternative to `btrfs`.

**Comment by jfmyers9** [comment]
@vito Is this quite a duplicate? 

Yes it is `btrfs` related, but it breaks the entire cluster. Concourse will not even schedule jobs on unaffected workers, workers with entirely different tags, and external workers (windows). That seems like more than just a `btrfs` issue. This has now occurred twice in one week since moving to 2.7.4.

**Comment by vito** [comment]
@jfmyers9 The issue says all workers were in the same state, i.e. btrfs hanging leaving processes in `D` state. Are you now saying that's not the case? Either way this issue as reported seems to be outlining specifically a hanging worker, and all indicators point to the same `btrfs` failure that we've seen in prod a few times ourselves. What we've seen is that a worker will be fine for a long time and then suddenly `btrfs` panics (you should see stuff in `/var/log/kernel` or wherever BOSH puts kernel logs) and the worker's disk is unrecoverable. We've never seen that break the entire cluster, though, and have never seen a pattern around when it happens.

I don't see any reason for this behavior to have changed in any of our 2.7.x releases. Maybe it's something with the stemcell?

If one worker failing in this way breaks the entire cluster that's worth investigating. A stack dump from the ATC would help. I don't see anything in the logs that would suggest a cluster failure; the `failed-to-get-volume-size` logs are benign and don't bubble up to anything else failing, and the rest don't seem related (or at least I have no way of knowing if they're talking about relevant workers). What I could see happening though is that a call is in-flight to the worker's BaggageClaim which is then hanging trying to do something with the disk. This will block forever and I wouldn't be surprised if that happens in some critical path to build scheduling (i.e. resource checking). We've already done the work to detect this failure state and recover from it in the story @joshzarrabi linked, so now workers will stall if their BaggageClaim is unresponsive, and requests on the client side will time out.

**Comment by jfmyers9** [comment]
I can see that "This affected all workers. (How is that even possible?) We have 4 types workers that are responsible for running different work loads. They are all exhibiting this behavior." may be somewhat misleading, but what was meant by that all the workers were unable to run jobs, not that they were suffering from the `btrfs` bug.

Yes this broke the entire cluster. We can wait for 2.7.5 to see if the above story fixes this issue. After looking at it yesterday, I agree that that seems like the most likely issue and matches the behavior that I was seeing in the logs. I also have `bosh logs` output from the `web` and single affected `worker` that broke the cluster in our most recent incident on Sunday, if there is somewhere that you would like me to upload them.
